File: src/pages/api/create-post-handler.ts
import type { APIRoute } from "astro";
import fs from "node:fs/promises";
import path from "node:path";
import yaml from "js-yaml"; // Import js-yaml
import { generateSlug } from "../../utils/slugify";
import type { PostApiPayload } from "../../types/admin";
import {
  transformApiPayloadToFrontmatter,
  generatePostFileContent,
} from "../../utils/adminApiHelpers";

export const prerender = false;

export const POST: APIRoute = async ({ request }) => {
  if (import.meta.env.PROD) {
    return new Response(
      JSON.stringify({ message: "Not available in production" }),
      {
        status: 403,
        headers: { "Content-Type": "application/json" },
      }
    );
  }

  try {
    const payload: PostApiPayload = await request.json();

    if (!payload.title || !payload.pubDate || !payload.postType) {
      return new Response(
        JSON.stringify({
          message: "Missing required fields (title, pubDate, postType)",
        }),
        {
          status: 400,
          headers: { "Content-Type": "application/json" },
        }
      );
    }

    const slug = generateSlug(payload.title || "untitled");
    const filename = `${slug}.mdx`; // Default to mdx for new posts
    const projectRoot = process.cwd();
    const filePath = path.join(projectRoot, "src", "content", "blog", filename);

    try {
      await fs.access(filePath);
      return new Response(
        JSON.stringify({
          message: `File already exists: ${filename}. Please use a different title.`,
        }),
        {
          status: 409, // Conflict
          headers: { "Content-Type": "application/json" },
        }
      );
    } catch { // _error removed
      // File does not exist, proceed with creation
    }

    const frontmatterObject = await transformApiPayloadToFrontmatter(payload); // Await the promise
    let generatedQuotesRef: string | undefined = undefined;

    // --- Handle Inline Quotes for New BookNote ---
    if (payload.postType === "bookNote") {
      generatedQuotesRef = `${slug}-quotes`;
      frontmatterObject.quotesRef = generatedQuotesRef; // Add to frontmatter BEFORE generating post content

      const quotesDir = path.join(projectRoot, "src", "content", "bookQuotes");
      const quotesFilePath = path.join(quotesDir, `${generatedQuotesRef}.yaml`);

      try {
        await fs.mkdir(quotesDir, { recursive: true });
    } catch (mkdirError: unknown) {
        console.error(
          `[API Create] Error creating bookQuotes directory ${quotesDir}:`,
        mkdirError instanceof Error ? mkdirError.message : String(mkdirError)
        );
        // Potentially return an error or log and continue without saving quotes
      }

      const quotesToSave = (payload.inlineQuotes || []).map((q) => ({
        text: q.text,
        quoteAuthor: q.quoteAuthor,
        tags: q.tags,
        quoteSource: q.quoteSource,
      }));

      const yamlData = {
        bookSlug: slug, // The slug of the post itself
        quotes: quotesToSave, // Will be an empty array if no inlineQuotes provided
      };

      try {
        const yamlString = yaml.dump(yamlData);
        await fs.writeFile(quotesFilePath, yamlString);
        if (import.meta.env.DEV) {
          console.log(
            `[API Create] Successfully created quotes file: ${quotesFilePath}`
          );
        }
      } catch (quoteError: unknown) {
        console.error(
          `[API Create] Error writing quotes file ${quotesFilePath}:`,
          quoteError instanceof Error ? quoteError.message : String(quoteError)
        );
        // Decide if this should be a critical error.
        // Could add a warning to the response message.
      }
    }
    // --- End Handle Inline Quotes ---

    const fileContent = generatePostFileContent(
      frontmatterObject,
      payload.bodyContent || "",
      payload.postType,
      true
    );

    await fs.writeFile(filePath, fileContent);

    interface ResponsePayload {
      message: string;
      filename: string;
      path: string;
      newSlug: string;
      title: string;
      quotesRef?: string;
    }

    const responsePayload: ResponsePayload = {
      message: "Post created successfully!",
      filename: filename,
      path: `/blog/${slug}`,
      newSlug: slug,
      title: frontmatterObject.title, // Return the processed title
    };

    if (generatedQuotesRef) {
      responsePayload.quotesRef = generatedQuotesRef;
    }

    return new Response(JSON.stringify(responsePayload), {
      status: 201,
      headers: { "Content-Type": "application/json" },
    });
  } catch (error: unknown) {
    let errorMessage = "An unknown error occurred.";
    if (error instanceof Error) {
      errorMessage = error.message;
    }

    if (
      error instanceof SyntaxError && // Check if error is SyntaxError
      errorMessage.toLowerCase().includes("json")
    ) {
      console.error("[API Create] Error parsing JSON body:", error);
      return new Response(
        JSON.stringify({
          message: "Invalid JSON data received.",
          errorDetail: errorMessage,
        }),
        {
          status: 400,
          headers: { "Content-Type": "application/json" },
        }
      );
    }
    console.error("[API Create] Error creating post:", error);
    return new Response(
      JSON.stringify({
        message: "Error creating post.",
        errorDetail: errorMessage,
      }),
      {
        status: 500,
        headers: { "Content-Type": "application/json" },
      }
    );
  }
};

File: go-backend/handler/create_post_handler.go
package handler

import (
	"fmt"
	"go-backend/model"
	"go-backend/utils"
	"net/http"
	"os"
	"path/filepath"

	"github.com/labstack/echo/v4"
	"gopkg.in/yaml.v3"
)

// CreatePostHandler handles POST requests to create new blog posts
// It mimics the behavior of the TypeScript create-post-handler.ts
func CreatePostHandler(c echo.Context) error {
	// Parse JSON payload
	var payload model.PostApiPayload
	if err := c.Bind(&payload); err != nil {
		return c.JSON(http.StatusBadRequest, model.ErrorResponse{
			Message:     "Invalid JSON data received.",
			ErrorDetail: err.Error(),
		})
	}

	// Validate required fields
	if payload.Title == "" || payload.PubDate == "" || payload.PostType == "" {
		return c.JSON(http.StatusBadRequest, model.ErrorResponse{
			Message: "Missing required fields (title, pubDate, postType)",
		})
	}

	// Generate slug from title
	slug := utils.GenerateSlug(payload.Title)
	if slug == "" {
		slug = "untitled"
	}

	// Create filename - default to .mdx for new posts
	filename := fmt.Sprintf("%s.mdx", slug)

	// Get project root (current working directory)
	projectRoot, err := os.Getwd()
	if err != nil {
		return c.JSON(http.StatusInternalServerError, model.ErrorResponse{
			Message:     "Error determining project root.",
			ErrorDetail: err.Error(),
		})
	}

	// Construct file path
	filePath := filepath.Join(projectRoot, "src", "content", "blog", filename)

	// Check if file already exists
	if _, err := os.Stat(filePath); err == nil {
		return c.JSON(http.StatusConflict, model.ErrorResponse{
			Message: fmt.Sprintf("File already exists: %s. Please use a different title.", filename),
		})
	}

	// Transform payload to frontmatter
	frontmatter, err := utils.TransformApiPayloadToFrontmatter(&payload)
	if err != nil {
		return c.JSON(http.StatusBadRequest, model.ErrorResponse{
			Message:     "Error processing post data.",
			ErrorDetail: err.Error(),
		})
	}

	var generatedQuotesRef string

	// Handle BookNote specific logic - create quotes file
	if payload.PostType == "bookNote" {
		generatedQuotesRef = fmt.Sprintf("%s-quotes", slug)
		frontmatter.QuotesRef = generatedQuotesRef

		// Create quotes directory
		quotesDir := filepath.Join(projectRoot, "src", "content", "bookQuotes")
		if err := os.MkdirAll(quotesDir, 0755); err != nil {
			return c.JSON(http.StatusInternalServerError, model.ErrorResponse{
				Message:     "Error creating bookQuotes directory.",
				ErrorDetail: err.Error(),
			})
		}

		// Prepare quotes to save
		quotesToSave := make([]model.Quote, 0, len(payload.InlineQuotes))
		for _, q := range payload.InlineQuotes {
			quotesToSave = append(quotesToSave, model.Quote{
				Text:        q.Text,
				QuoteAuthor: q.QuoteAuthor,
				Tags:        q.Tags,
				QuoteSource: q.QuoteSource,
			})
		}

		// Create quotes YAML structure
		quotesYAML := model.QuotesYAML{
			BookSlug: slug,
			Quotes:   quotesToSave,
		}

		// Write quotes file
		quotesFilePath := filepath.Join(quotesDir, fmt.Sprintf("%s.yaml", generatedQuotesRef))
		quotesYAMLBytes, err := yaml.Marshal(quotesYAML)
		if err != nil {
			return c.JSON(http.StatusInternalServerError, model.ErrorResponse{
				Message:     "Error marshaling quotes to YAML.",
				ErrorDetail: err.Error(),
			})
		}

		if err := os.WriteFile(quotesFilePath, quotesYAMLBytes, 0644); err != nil {
			return c.JSON(http.StatusInternalServerError, model.ErrorResponse{
				Message:     "Error writing quotes file.",
				ErrorDetail: err.Error(),
			})
		}
	}

	// Generate post file content
	fileContent, err := utils.GeneratePostFileContent(frontmatter, payload.BodyContent, payload.PostType, true)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, model.ErrorResponse{
			Message:     "Error generating post content.",
			ErrorDetail: err.Error(),
		})
	}

	// Ensure the blog directory exists
	blogDir := filepath.Join(projectRoot, "src", "content", "blog")
	if err := os.MkdirAll(blogDir, 0755); err != nil {
		return c.JSON(http.StatusInternalServerError, model.ErrorResponse{
			Message:     "Error creating blog directory.",
			ErrorDetail: err.Error(),
		})
	}

	// Write post file
	if err := os.WriteFile(filePath, []byte(fileContent), 0644); err != nil {
		return c.JSON(http.StatusInternalServerError, model.ErrorResponse{
			Message:     "Error writing post file.",
			ErrorDetail: err.Error(),
		})
	}

	// Prepare success response
	response := model.SuccessResponse{
		Message:  "Post created successfully!",
		Filename: filename,
		Path:     fmt.Sprintf("/blog/%s", slug),
		NewSlug:  slug,
		Title:    frontmatter.Title,
	}

	if generatedQuotesRef != "" {
		response.QuotesRef = generatedQuotesRef
	}

	return c.JSON(http.StatusCreated, response)
}

// SetupRoutes configures the Echo routes for the post creation API
func SetupRoutes(e *echo.Echo) {
	// API routes
	api := e.Group("/api")
	api.POST("/create-post", CreatePostHandler)
}

// CORS middleware for development
func CORSMiddleware() echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			c.Response().Header().Set("Access-Control-Allow-Origin", "*")
			c.Response().Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
			c.Response().Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization")

			if c.Request().Method == "OPTIONS" {
				return c.NoContent(http.StatusNoContent)
			}

			return next(c)
		}
	}
}
